### ISSUE TITLE
`push_to_git_remote` failed on a Mac that had pulled tags from remote

### ISSUE THREAD

**Issue by mokagio** [wordpress-mobile/release-toolkit#314]
This is the error I got:

```
/Users/gio/Developer/a8c/wpios/vendor/bundle/ruby/2.7.0/gems/fastlane-2.195.0/fastlane_core/lib/fastlane_core/ui/interface.rb:153:in `shell_error!': \e[0;31;49m[!] Exit status of command 'git push origin release/18.8:release/18.8 --tags' was 1 instead of 0. (FastlaneCore::Interface::FastlaneShellError)
remote: This repository moved. Please use the new location:
remote:   git@github.com:wordpress-mobile/WordPress-iOS.git
To github.com:wordpress-mobile/wordpress-ios
   ce978a3267..4251fe5a6a  release/18.8 -> release/18.8
 ! [rejected]              16.6 -> 16.6 (already exists)
error: failed to push some refs to 'github.com:wordpress-mobile/wordpress-ios'
```

@AliSoftware I think it should be straightforward to tweak the push command to avoid this (remove the `--tags`? 🤔 ). But, I run out of time today so dumping this here without any research is the best I could do 😅

**Comment by AliSoftware** [comment]
@mokagio Do you know if this could have happened only if the local vs remote tag didn't match (i.e. didn't point to the same commit)?

I don't think even a pristine `git push --tags` in the terminal would generate such an error if the remote tag already exists… except if tags didn't match and were out of phase. If fact, I can usually run `git push --tags` on any repo as many times as I want and it won't complain that the tags already exists on remote even after the 1st run… So the issue is not really that it tries to push a tag that already exists, but rather than there's a mismatch with your local repo.

Can you check that your local tags (especially `16.6`) all match the remote ones? Or, alternatively, do a `git pull --tags --force` to ensure your local working copy and tags are sync'd with the remotes (I think that's the right command for that?)… and then try to run `git push origin release/18.8:release/18.8 --tags` in your Terminal again?

---

I just did the test on my end, by running this command in my terminal multiple times, without any failure about tags already existing, so I think that's still a genuine error you had.

```
 18:25:50  📂 ~/Documents/Dev/apps/WordPress-iOS/  (release/18.8) 
✓ $ git push origin release/18.8:release/18.8 --tags

Everything up-to-date
                                                                                                                                                                                                                
 18:26:02  📂 ~/Documents/Dev/apps/WordPress-iOS/  (release/18.8) 
✓ $ git push origin release/18.8:release/18.8 --tags

Everything up-to-date
```

That being said, it would make sense not to push tags when we don't really _need_ to do so at that step in our process. This would be a simple-enough fix, since the fastlane built-in `push_to_git_remote` action includes a `tags:` parameter (true by default) so we'd just have to set it to `tags: false` at call site in the `Fastfile`.

That would make sense to do so for that particular lane and process step, so I'll try to make that change in WPiOS's `Fastfile` at some point; though that still doesn't change the fact that it seems that your local working copy is out of sync with the remote tags and that it would be a good thing to fix that on your Mac by re-synching your local tags with remote at some point 🙂

**Comment by AliSoftware** [comment]
Should be fixed via https://github.com/wordpress-mobile/WordPress-iOS/pull/17627 to reduce that risk

I still recommend that you fix your local git working copy of WPiOS though to resync your tags 😜

**Comment by mokagio** [comment]
Thanks for following up on this @AliSoftware 👍 

My turn, now.

> @mokagio Do you know if this could have happened only if the local vs remote tag didn't match (i.e. didn't point to the same commit)?

`16.6` seems to match, or, at least, point to the same [commit](https://github.com/wordpress-mobile/WordPress-iOS/commit/40e66ade46e6525203e97665abb4b7b7a9aef325):

<img width="1791" alt="Screen Shot 2021-12-06 at 7 56 18 pm" src="https://user-images.githubusercontent.com/1218433/144906305-2c47283e-fb1b-4db0-97b7-60bb61daecf2.png">

Interestingly, some other tags on my local copy seem to differ, somehow, but not `16.6` 🤔 :

```
➜ git pull --tags
From github.com:wordpress-mobile/WordPress-iOS
 ! [rejected]              15.4       -> 15.4  (would clobber existing tag)
 ! [rejected]              16.0       -> 16.0  (would clobber existing tag)
 ! [rejected]              17.5.0.1   -> 17.5.0.1  (would clobber existing tag)
```

I also tried `git fetch --tags`, which gave me the same error:

```
➜ gf --tags
From github.com:wordpress-mobile/WordPress-iOS
 - [deleted]               (none)     -> origin/issue/17503-change-username
remote: Enumerating objects: 88218, done.
remote: Counting objects: 100% (4100/4100), done.
remote: Compressing objects: 100% (23/23), done.
remote: Total 88218 (delta 4077), reused 4095 (delta 4077), pack-reused 84118
Receiving objects: 100% (88218/88218), 12.55 MiB | 2.91 MiB/s, done.
Resolving deltas: 100% (57278/57278), completed with 207 local objects.
   92531a10ab..a5a06384c5  develop                          -> origin/develop
 * [new branch]            issue/17476-moderation-actions   -> origin/issue/17476-moderation-actions
 * [new branch]            tooling/fix-push                 -> origin/tooling/fix-push
 * [new branch]            tooling/l10n-step2/ios13-widgets -> origin/tooling/l10n-step2/ios13-widgets
 * [new tag]               10.44                            -> 10.44
 * [new tag]               10.44.0.0                        -> 10.44.0.0
 * [new tag]               11.0.0.0                         -> 11.0.0.0
 * [new tag]               11.0.0.20181010                  -> 11.0.0.20181010
 * [new tag]               11.1.0.0                         -> 11.1.0.0
 * [new tag]               11.1.0.20181023                  -> 11.1.0.20181023
 * [new tag]               12.1.1.0                         -> 12.1.1.0
 * [new tag]               12.1.1.20190411                  -> 12.1.1.20190411
 * [new tag]               12.1.2.0                         -> 12.1.2.0
 * [new tag]               12.1.2.20190411                  -> 12.1.2.20190411
 * [new tag]               12.3.0.2                         -> 12.3.0.2
 * [new tag]               12.4.0.0                         -> 12.4.0.0
 * [new tag]               12.4.0.20190423                  -> 12.4.0.20190423
 * [new tag]               12.5.0.0                         -> 12.5.0.0
 * [new tag]               12.5.0.20190520                  -> 12.5.0.20190520
 * [new tag]               12.6.0.20190619                  -> 12.6.0.20190619
 * [new tag]               12.6.0.6                         -> 12.6.0.6
 * [new tag]               14.3.1                           -> 14.3.1
 * [new tag]               14.3.1.0                         -> 14.3.1.0
 * [new tag]               14.44                            -> 14.44
 ! [rejected]              15.4                             -> 15.4  (would clobber existing tag)
 ! [rejected]              16.0                             -> 16.0  (would clobber existing tag)
 * [new tag]               16.8.1                           -> 16.8.1
 ! [rejected]              17.5.0.1                         -> 17.5.0.1  (would clobber existing tag)
```

I then tried to `git push --tags` to see if it failed with the same `16.6` error, and realized I re-pushed some of the old intermediate tags  to `origin` 😕 

```
➜ gp --tags
Total 0 (delta 0), reused 0 (delta 0), pack-reused 0
To github.com:wordpress-mobile/WordPress-iOS
 * [new tag]               14.8.0.0 -> 14.8.0.0
 * [new tag]               14.8.0.1 -> 14.8.0.1
 * [new tag]               14.8.0.2 -> 14.8.0.2
 * [new tag]               14.8.0.20200504 -> 14.8.0.20200504
 * [new tag]               14.8.0.20200508 -> 14.8.0.20200508
 * [new tag]               14.8.0.20200513 -> 14.8.0.20200513
 * [new tag]               14.9.0.0 -> 14.9.0.0
 * [new tag]               14.9.0.1 -> 14.9.0.1
 * [new tag]               14.9.0.2 -> 14.9.0.2
 * [new tag]               14.9.0.20200518 -> 14.9.0.20200518
 * [new tag]               14.9.0.20200519 -> 14.9.0.20200519
 * [new tag]               14.9.0.20200522 -> 14.9.0.20200522
 * [new tag]               15.1.0.0 -> 15.1.0.0
 * [new tag]               15.1.0.1 -> 15.1.0.1
 * [new tag]               15.1.0.2 -> 15.1.0.2
 * [new tag]               15.1.0.20200615 -> 15.1.0.20200615
 * [new tag]               15.1.0.20200616 -> 15.1.0.20200616
 * [new tag]               15.1.0.20200622 -> 15.1.0.20200622
 * [new tag]               15.3.0.0 -> 15.3.0.0
 * [new tag]               15.3.0.1 -> 15.3.0.1
 * [new tag]               15.3.0.2 -> 15.3.0.2
 * [new tag]               15.3.0.20200713 -> 15.3.0.20200713
 * [new tag]               15.3.0.20200720 -> 15.3.0.20200720
 * [new tag]               15.3.0.20200722 -> 15.3.0.20200722
 * [new tag]               15.4.0.0 -> 15.4.0.0
 * [new tag]               15.4.0.1 -> 15.4.0.1
 * [new tag]               15.4.0.2 -> 15.4.0.2
 * [new tag]               15.4.0.20200727 -> 15.4.0.20200727
 * [new tag]               15.4.0.20200730 -> 15.4.0.20200730
 * [new tag]               15.4.0.20200804 -> 15.4.0.20200804
 * [new tag]               15.5.0.0 -> 15.5.0.0
 * [new tag]               15.5.0.1 -> 15.5.0.1
 * [new tag]               15.5.0.2 -> 15.5.0.2
 * [new tag]               15.5.0.20200810 -> 15.5.0.20200810
 * [new tag]               15.5.0.20200814 -> 15.5.0.20200814
 * [new tag]               15.5.0.20200818 -> 15.5.0.20200818
 * [new tag]               15.6.0.0 -> 15.6.0.0
 * [new tag]               15.6.0.1 -> 15.6.0.1
 * [new tag]               15.6.0.2 -> 15.6.0.2
 * [new tag]               15.6.0.20200824 -> 15.6.0.20200824
 * [new tag]               15.6.0.20200825 -> 15.6.0.20200825
 * [new tag]               15.6.0.20200828 -> 15.6.0.20200828
 * [new tag]               15.7.0.0 -> 15.7.0.0
 * [new tag]               15.7.0.1 -> 15.7.0.1
 * [new tag]               15.7.0.20200907 -> 15.7.0.20200907
 * [new tag]               15.7.0.20200909 -> 15.7.0.20200909
 ! [rejected]              15.4 -> 15.4 (already exists)
 ! [rejected]              16.0 -> 16.0 (already exists)
 ! [rejected]              17.5.0.1 -> 17.5.0.1 (already exists)
error: failed to push some refs to 'github.com:wordpress-mobile/WordPress-iOS'
hint: Updates were rejected because the tag already exists in the remote.
```

Three things I'm curious about:

- Why I have those different tags? 
- Why I got an error on `16.6` that day but not today? _At first, I thought it was a typo, but that text is clearly copy pasted from my terminal_
- Why that doesn't happen to you?

I remember there being some changes to tags in the past. Before I dive into a Git spelunking (or don't do it in favor of more impactful work while feeling bad about leaving an open loop here) I'll summon @jkmassel: does this ring any bell? 

For the record, I'm running on Git `2.33.0` and [my `push` strategy is `simple`](https://github.com/mokagio/dotfiles/blob/87ea9d2f404ba5b903e9d604affe1b47a39099ba/gitconfig#L6-L7).

**Comment by AliSoftware** [comment]
> Why I have those different tags?

The only reason I can think for the "already exists" or "would clobber existing tag" errors if if you pulled the tags at some point, then they later got moved to a different commit on the remote, or got deleted and recreated and someone git consider them different.

Though as you said, `git show` clearly showed you that your locall `16.6` pointed to the same commit as the remote one, so this is strange… maybe another diff that could have happened that makes git consider them different is [if they are annotated tags](https://git-scm.com/book/en/v2/Git-Basics-Tagging#:~:text=list%20is%20mandatory.-,Creating%20Tags,-Git%20supports%20two) (or your local tag was annotated, but the remote was recreated and a lightweight tag… or with a different annotation…)?!

Another explanation would be that you kinda accidentally updated/fixed the 16.6 tag since your last test, i.e. back when you did the code_freeze and had that issue, your 16.6 tags indeed were out of sync… but somewhere later down the road (maybe hidden under some lane executing a `pull --tags --force`, or your Git GUI – if you use one – doing it for you in the background…?) it got resync'd without you knowing, which explains why it's now fixed, why `git show` now points to the same commit for both local and remote tags, and why you don't have the error on 16.6 today anymore? Still feels like a stretch, but at that point I don't have many more ideas…

> Why that doesn't happen to you?

This being all about local tags being out of sync, I'm not that surprised it happened to you but not me, because you being the WPiOS release manager, you manipulate the WPiOS repo way more often than me, and you're the one running the release lanes locally, and `git pull`-ing the repo on a more regular basis. Which I guess makes it more likely for you to pull any new commit/tags both before and after they might change remotely (why they'd change is a mystery, but if they do, you'd still be more likely to have already pulled the tag before the move).
While I personally only work on WPiOS repo on rare occasions, so when I get back to it and do a `git pull` (well, more often [a `git groom`](https://gist.github.com/AliSoftware/be7fd7de9468adaf4e47b615bbecfb1c) in practice), way later and after those tags have been created, so they had time to maybe accidentally be moved in the past (when I never pulled the intermediate states of those tags) but a couple of sprints later once I pull them for the first time on my local mac they are now stable…

That still doesn't explain why those tags would have moved in the first place or differ between your local and the remote, though I know that a while ago – especially back when we created the tags locally via Fastlane before pushing to CI, before we migrated to making the GitHub Release publication create the tag like we now do – we used to create those tags locally which made them more prone to be deleted and recreated (if the CI failed building the tag we pushed and we had to fix it and move the tag to the new commit). I can't remember at which version we did that change of process tho, but that could be the best explanation for the cases of tags `15.4`, `16.0` and `17.5.0.1 `?

**Comment by AliSoftware** [comment]
@mokagio PS: Notice that my `git groom` script that I use very often especially when I get back to a repo I haven't been working on for a while, to ensure I get back to a more up-to-date state by pulling all the recent changes in the main branches and pruning all old ones… [also happens to prune any orphan tags that could have been deleted](https://gist.github.com/AliSoftware/be7fd7de9468adaf4e47b615bbecfb1c#file-git-groom-L56). Maybe that contributed to helping me having my local repo more-up-to-date with the remote with less chances of having old-and-since-moved-or-deleted tags staying around in my local working copy?
